File: task5/main.go
package main

import "fmt"

type Pistol struct {
	model            string
	caliber          string
	magazineCapacity int
	roundsInMagazine int
}

func NewPistol(model, caliber string, magazineCapacity int) Pistol {
	return Pistol{
		model:            model,
		caliber:          caliber,
		magazineCapacity: magazineCapacity,
		roundsInMagazine: 0,
	}
}

func (p *Pistol) Info() string {
	return fmt.Sprintf("Model: %s, Caliber: %s, Capacity: %d, Rounds: %d",
		p.model, p.caliber, p.magazineCapacity, p.roundsInMagazine)
}

func (p *Pistol) Reload(rounds int) {
	if rounds < 0 {
		rounds = 0
	}
	if rounds > p.magazineCapacity {
		rounds = p.magazineCapacity
	}
	p.roundsInMagazine = rounds
}

func (p *Pistol) Shoot() bool {
	if p.roundsInMagazine == 0 {
		return false
	}
	p.roundsInMagazine--
	return true
}

func main() {
	pistol := NewPistol("Glock 17", "9mm", 17)

	fmt.Println(pistol.Info())

	pistol.Reload(17)
	fmt.Println("After reload:", pistol.Info())

	for i := 1; i <= 3; i++ {
		if pistol.Shoot() {
			fmt.Printf("Shot %d fired. %s\n", i, pistol.Info())
		}
	}
}

File: task5/main.ts
class Pistol {
  constructor(
    private model: string,
    private caliber: string,
    private magazineCapacity: number,
    private roundsInMagazine: number = 0
  ) {}

  static create(
    model: string,
    caliber: string,
    magazineCapacity: number
  ): Pistol {
    return new Pistol(model, caliber, magazineCapacity);
  }

  info(): string {
    return `Model: ${this.model}, Caliber: ${this.caliber}, Capacity: ${this.magazineCapacity}, Rounds: ${this.roundsInMagazine}`;
  }

  reload(rounds: number): void {
    const loaded = Math.max(0, Math.min(rounds, this.magazineCapacity));
    this.roundsInMagazine = loaded;
  }

  shoot(): boolean {
    if (this.roundsInMagazine === 0) {
      return false;
    }
    this.roundsInMagazine--;
    return true;
  }
}

const pistol = Pistol.create("Glock 17", "9mm", 17);

console.log(pistol.info());

pistol.reload(17);
console.log("After reload:", pistol.info());

for (let i = 1; i <= 3; i++) {
  if (pistol.shoot()) {
    console.log(`Shot ${i} fired. ${pistol.info()}`);
  }
}
